lib/BloqadeSchema/src/transform.jl

```diff
@@ -0,0 +1,389 @@
+# function find_local_masks(values::Array{T,2};name::Symbol=:Waveform,ntrunc::Int=1, assert_truncation::Bool=false) where T
+    
+#     (l,w) = size(values)
+    
+#     # project out uniform amplitude waveforms
+#     ones_norm = ones(w,1)./sqrt(w)
+#     avg = (values*ones_norm)*transpose(ones_norm)
+#     results = []
+#     values = values - avg
+#     avg = avg[:,1]
+
+#     if ntrunc > 0
+#         # use SVD to decompose the non-uniform waveforms
+#         # if there are more than one singular value larger than 0 the remaining pattern
+#         # can't be described as a local detuning mask times a single time-dependent function.
+#         # here we truncate the waveform only keeping the first singular value.
+#         u,s,v_T = svd(values)
+#         contained_weight = sum(s[1:ntrunc])
+        
+#         remaining_weight = sum(s[(ntrunc+1):end])
+
+#         if remaining_weight > eps()*length(s)*contained_weight
+#             error_or_warn(assert_truncation,"Cannot decompose $name into product of masks and scalar functions.")
+#         end
+
+#         for i in 1:ntrunc
+
+#             mask = if all(u[:,i] .<= 0) # make u all positive
+#                 u[:,i] .*= -1
+#                 (-s[i] .* v_T[:,i])
+
+#             else
+#                 s[i] .* v_T[:,i]
+#             end
+
+#             min = minimum(mask)
+#             max = maximum(mask)
+
+#             mask = (mask .- min) ./ (max - min)
+
+#             f = u[:,i] .* (max - min)
+#             avg .+= u[:,i] .* min
+
+#             push!(results,(f,mask))
+#         end
+#     end
+
+#     push!(results,(avg,ones(w)))
+#     return results
+
+# end
+function clip_waveform(wf::Waveform{BloqadeWaveforms.PiecewiseLinear{T,I},T},name,min_value::T,max_value::T) where {T<:Real,I}
+    @assert min_value < max_value
+
+    for value in wf.f.values
+        if value > max_value || value < min_value
+            @info "During hardware transform: $name(t) falls outside of hardware bounds, clipping to maximum/minimum."
+            break
+        end
+    end
+
+    return piecewise_linear(;clocks=wf.f.clocks,values=map(wf.f.values) do value
+            return min(max_value,max(min_value,value))
+        end
+    )
+end
+
+function hardware_transform_Ω(Ω,device_capabilities::DeviceCapabilities)
+    time_res = device_capabilities.rydberg.global_value.timeResolution
+    min_step = device_capabilities.rydberg.global_value.timeDeltaMin
+    rabi_res = device_capabilities.rydberg.global_value.rabiFrequencyResolution
+    max_slope = device_capabilities.rydberg.global_value.rabiFrequencySlewRateMax
+    max_value = device_capabilities.rydberg.global_value.rabiFrequencyMax
+    min_value = device_capabilities.rydberg.global_value.rabiFrequencyMin
+
+    check_waveform(Ω,:Ω)
+    check_global(Ω,:Ω)
+    warn_duration(time_res,Ω,:Ω)
+
+    Ω = if !isapprox(Ω(0.0), 0.0;atol=eps(),rtol=√eps()) || !isapprox(Ω(Ω.duration),0.0;atol=eps(),rtol=√eps())
```

Comment:
remove this `if` statement, this should be captured by the `pin_waveform_edges` function.

Reply:
`\Omega = pin_waveform_edges(\Omega,max_slope,0,0)` @johnzl-777